src/connector/src/sink/sqlserver.rs

```diff
@@ -0,0 +1,625 @@
+                        self.config.table,
+                        all_col_names,
+                        param_placeholders(&mut next_param_id),
+                    )
+                    .unwrap();
+                }
+                SqlOp::Merge(_) => {
+                    write!(
+                        &mut query_str,
+                        r#"MERGE {} AS TARGET
+                        USING (VALUES ({})) AS SOURCE ({})
+                        ON ({})
+                        WHEN MATCHED THEN UPDATE SET {}
+                        WHEN NOT MATCHED THEN INSERT ({}) VALUES ({});"#,
+                        self.config.table,
+                        param_placeholders(&mut next_param_id),
+                        all_col_names,
+                        pk_match,
+                        set_all_source_col,
+                        all_col_names,
+                        all_source_col_names,
+                    )
+                    .unwrap();
+                }
+                SqlOp::Delete(_) => {
+                    write!(
+                        &mut query_str,
+                        r#"DELETE FROM {} WHERE {}"#,
+                        self.config.table,
+                        self.pk_indices
+                            .iter()
+                            .map(|idx| {
+                                let condition =
+                                    format!("{}=@P{}", self.schema[*idx].name, next_param_id);
+                                next_param_id += 1;
+                                condition
+                            })
+                            .collect::<Vec<_>>()
+                            .join(","),
+                    )
+                    .unwrap();
+                }
+            }
+        }
+        tracing::debug!(query_str, "!!!");
+
+        let mut query = Query::new(query_str);
+        for op in self.ops.drain(..) {
+            match op {
+                SqlOp::Insert(row) => {
+                    bind_params(&mut query, row, &self.schema, 0..col_num)?;
+                }
+                SqlOp::Merge(row) => {
+                    bind_params(&mut query, row, &self.schema, 0..col_num)?;
+                }
+                SqlOp::Delete(row) => {
+                    bind_params(
+                        &mut query,
+                        row,
+                        &self.schema,
+                        self.pk_indices.iter().copied(),
+                    )?;
+                }
+            }
+        }
+        query.execute(&mut self.sql_client.client).await?;
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl SinkWriter for SqlServerSinkWriter {
+    async fn begin_epoch(&mut self, _epoch: u64) -> Result<()> {
+        Ok(())
+    }
+
+    async fn write_batch(&mut self, chunk: StreamChunk) -> Result<()> {
+        for (op, row) in chunk.rows() {
+            match op {
+                Op::Insert => {
+                    self.upsert_one(row);
```

Comment:
`insert_one` is useless?

Reply:
Could you clarify the expected behavior when sink is append-only, downstream has PK, should RisingWave insert or  upsert to downstream?

Reply:
For append-only sink user doesn't need to specify the `primary_key` property, so the downstream table should not define a primary key (we can have a check for it) then we can use `INSERT` to write out the data.

Reply:
Made Op::Insert for append-only and Op::Upsert for non-append-only.